// ngx-color/slider/esm5/slider-swatch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
var SliderSwatchComponent = /** @class */ (function () {
    function SliderSwatchComponent() {
        this.first = false;
        this.last = false;
        this.onClick = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SliderSwatchComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.background = "hsl(" + this.hsl.h + ", 50%, " + this.offset * 100 + "%)";
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SliderSwatchComponent.prototype.handleClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.onClick.emit({
            data: {
                h: this.hsl.h,
                s: 0.5,
                l: this.offset,
                source: 'hsl',
            },
            $event: $event,
        });
    };
    SliderSwatchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-slider-swatch',
                    template: "\n  <div class=\"slider-swatch\" [style.background]=\"background\"\n    [class.first]=\"first\" [class.last]=\"last\" [class.active]=\"active\"\n    (click)=\"handleClick($event)\"\n  ></div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n    .slider-swatch {\n      height: 12px;\n      background: rgb(121, 211, 166);\n      cursor: pointer;\n    }\n    .slider-swatch.active {\n      transform: scaleY(1.8);\n      border-top-right-radius: 3.6px 2px;\n      border-top-left-radius: 3.6px 2px;\n      border-bottom-right-radius: 3.6px 2px;\n      border-bottom-left-radius: 3.6px 2px;\n    }\n    .slider-swatch.first {\n      border-radius: 2px 0px 0px 2px;\n    }\n    .slider-swatch.last {\n      border-radius: 0px 2px 2px 0px;\n    }\n\n  "]
                }] }
    ];
    SliderSwatchComponent.propDecorators = {
        hsl: [{ type: Input }],
        active: [{ type: Input }],
        offset: [{ type: Input }],
        first: [{ type: Input }],
        last: [{ type: Input }],
        onClick: [{ type: Output }]
    };
    return SliderSwatchComponent;
}());
export { SliderSwatchComponent };
if (false) {
    /** @type {?} */
    SliderSwatchComponent.prototype.hsl;
    /** @type {?} */
    SliderSwatchComponent.prototype.active;
    /** @type {?} */
    SliderSwatchComponent.prototype.offset;
    /** @type {?} */
    SliderSwatchComponent.prototype.first;
    /** @type {?} */
    SliderSwatchComponent.prototype.last;
    /** @type {?} */
    SliderSwatchComponent.prototype.onClick;
    /** @type {?} */
    SliderSwatchComponent.prototype.background;
}
