// ngx-color/circle/bundles/ngx-color-circle.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common'), require('material-colors'), require('ngx-color')) :
    typeof define === 'function' && define.amd ? define('ngx-color/circle', ['exports', '@angular/core', '@angular/common', 'material-colors', 'ngx-color'], factory) :
    (factory((global['ngx-color'] = global['ngx-color'] || {}, global['ngx-color'].circle = {}),global.ng.core,global.ng.common,global['material-colors'],global['ngx-color']));
}(this, (function (exports,core,common,materialColors,ngxColor) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var CircleSwatchComponent = /** @class */ (function () {
        function CircleSwatchComponent() {
            this.circleSize = 28;
            this.circleSpacing = 14;
            this.focus = false;
            this.onClick = new core.EventEmitter();
            this.onSwatchHover = new core.EventEmitter();
            this.swatchStyle = {
                borderRadius: '50%',
                background: 'transparent',
                transition: '100ms box-shadow ease',
            };
        }
        /**
         * @return {?}
         */
        CircleSwatchComponent.prototype.ngOnChanges = /**
         * @return {?}
         */
            function () {
                this.focusStyle = {
                    boxShadow: this.color + " 0px 0px 0px 3px inset",
                };
                this.swatchStyle.boxShadow = "inset 0 0 0 " + this.circleSize / 2 + "px " + this.color;
            };
        /**
         * @param {?} __0
         * @return {?}
         */
        CircleSwatchComponent.prototype.handleClick = /**
         * @param {?} __0
         * @return {?}
         */
            function (_a) {
                var hex = _a.hex, $event = _a.$event;
                this.onClick.emit({ hex: hex, $event: $event });
            };
        CircleSwatchComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'color-circle-swatch',
                        template: "\n  <div class=\"circle-swatch\"\n    [style.width.px]=\"circleSize\" [style.height.px]=\"circleSize\"\n    [style.margin-right.px]=\"circleSpacing\" [style.margin-bottom.px]=\"circleSpacing\"\n    >\n    <color-swatch\n      [color]=\"color\" [style]=\"swatchStyle\" [focus]=\"focus\" [focusStyle]=\"focusStyle\"\n      (onClick)=\"handleClick($event)\" (onHover)=\"onSwatchHover.emit($event)\">\n    </color-swatch>\n    <div class=\"clear\"></div>\n  </div>\n  ",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        preserveWhitespaces: false,
                        styles: ["\n  .circle-swatch {\n    transform: scale(1);\n    transition: transform 100ms ease;\n  }\n  .circle-swatch:hover {\n    transform: scale(1.2);\n  }\n  "]
                    }] }
        ];
        CircleSwatchComponent.propDecorators = {
            color: [{ type: core.Input }],
            circleSize: [{ type: core.Input }],
            circleSpacing: [{ type: core.Input }],
            focus: [{ type: core.Input }],
            onClick: [{ type: core.Output }],
            onSwatchHover: [{ type: core.Output }]
        };
        return CircleSwatchComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var CircleComponent = /** @class */ (function (_super) {
        __extends(CircleComponent, _super);
        function CircleComponent() {
            var _this = _super.call(this) || this;
            /**
             * Pixel value for picker width
             */
            _this.width = 252;
            /**
             * Color squares to display
             */
            _this.colors = [
                materialColors.red['500'],
                materialColors.pink['500'],
                materialColors.purple['500'],
                materialColors.deepPurple['500'],
                materialColors.indigo['500'],
                materialColors.blue['500'],
                materialColors.lightBlue['500'],
                materialColors.cyan['500'],
                materialColors.teal['500'],
                materialColors.green['500'],
                materialColors.lightGreen['500'],
                materialColors.lime['500'],
                materialColors.yellow['500'],
                materialColors.amber['500'],
                materialColors.orange['500'],
                materialColors.deepOrange['500'],
                materialColors.brown['500'],
                materialColors.blueGrey['500'],
            ];
            /**
             * Value for circle size
             */
            _this.circleSize = 28;
            /**
             * Value for spacing between circles
             */
            _this.circleSpacing = 14;
            return _this;
        }
        /**
         * @param {?} color
         * @return {?}
         */
        CircleComponent.prototype.isActive = /**
         * @param {?} color
         * @return {?}
         */
            function (color) {
                return this.hex === color;
            };
        /**
         * @param {?} __0
         * @return {?}
         */
        CircleComponent.prototype.handleBlockChange = /**
         * @param {?} __0
         * @return {?}
         */
            function (_a) {
                var hex = _a.hex, $event = _a.$event;
                if (ngxColor.isValidHex(hex)) {
                    this.handleChange({ hex: hex, source: 'hex' }, $event);
                }
            };
        /**
         * @param {?} __0
         * @return {?}
         */
        CircleComponent.prototype.handleValueChange = /**
         * @param {?} __0
         * @return {?}
         */
            function (_a) {
                var data = _a.data, $event = _a.$event;
                this.handleChange(data, $event);
            };
        CircleComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'color-circle',
                        template: "\n  <div class=\"circle-picker {{ className }}\"\n    [style.width.px]=\"width\"\n    [style.margin-right.px]=\"-circleSpacing\"\n    [style.margin-bottom.px]=\"-circleSpacing\">\n    <color-circle-swatch\n      *ngFor=\"let color of colors\"\n      [color]=\"color\" [focus]=\"isActive(color)\"\n      (onClick)=\"handleBlockChange($event)\"\n      (onSwatchHover)=\"onSwatchHover.emit($event)\"\n    ></color-circle-swatch>\n  </div>\n  ",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        preserveWhitespaces: false,
                        styles: ["\n      .circle-picker {\n        display: flex;\n        flex-wrap: wrap;\n      }\n    "]
                    }] }
        ];
        /** @nocollapse */
        CircleComponent.ctorParameters = function () { return []; };
        CircleComponent.propDecorators = {
            width: [{ type: core.Input }],
            colors: [{ type: core.Input }],
            circleSize: [{ type: core.Input }],
            circleSpacing: [{ type: core.Input }]
        };
        return CircleComponent;
    }(ngxColor.ColorWrap));
    var ColorCircleModule = /** @class */ (function () {
        function ColorCircleModule() {
        }
        ColorCircleModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [CircleComponent, CircleSwatchComponent],
                        exports: [CircleComponent, CircleSwatchComponent],
                        imports: [common.CommonModule, ngxColor.SwatchModule],
                    },] }
        ];
        return ColorCircleModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.CircleComponent = CircleComponent;
    exports.ColorCircleModule = ColorCircleModule;
    exports.ɵa = CircleSwatchComponent;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
