// ngx-color/alpha/bundles/ngx-color-alpha.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common'), require('@angular/core'), require('ngx-color')) :
    typeof define === 'function' && define.amd ? define('ngx-color/alpha', ['exports', '@angular/common', '@angular/core', 'ngx-color'], factory) :
    (factory((global['ngx-color'] = global['ngx-color'] || {}, global['ngx-color'].alpha = {}),global.ng.common,global.ng.core,global['ngx-color']));
}(this, (function (exports,common,core,ngxColor) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var AlphaPickerComponent = /** @class */ (function (_super) {
        __extends(AlphaPickerComponent, _super);
        function AlphaPickerComponent() {
            var _this = _super.call(this) || this;
            /**
             * Pixel value for picker width
             */
            _this.width = 316;
            /**
             * Pixel value for picker height
             */
            _this.height = 16;
            _this.direction = 'horizontal';
            _this.pointer = {
                width: '18px',
                height: '18px',
                borderRadius: '50%',
                transform: 'translate(-9px, -2px)',
                boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
            };
            return _this;
        }
        /**
         * @return {?}
         */
        AlphaPickerComponent.prototype.ngOnChanges = /**
         * @return {?}
         */
            function () {
                if (this.direction === 'vertical') {
                    this.pointer.transform = 'translate(-3px, -9px)';
                }
                this.setState(ngxColor.toState(this.color, this.oldHue));
            };
        /**
         * @param {?} __0
         * @return {?}
         */
        AlphaPickerComponent.prototype.handlePickerChange = /**
         * @param {?} __0
         * @return {?}
         */
            function (_a) {
                var data = _a.data, $event = _a.$event;
                this.handleChange(data, $event);
            };
        AlphaPickerComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'color-alpha-picker',
                        template: "\n    <div class=\"alpha-picker {{ className }}\"\n      [style.width.px]=\"width\" [style.height.px]=\"height\">\n      <color-alpha\n        [hsl]=\"hsl\"\n        [rgb]=\"rgb\"\n        [pointer]=\"pointer\"\n        [direction]=\"direction\"\n        (onChange)=\"handlePickerChange($event)\"\n      ></color-alpha>\n    </div>\n  ",
                        changeDetection: core.ChangeDetectionStrategy.OnPush,
                        preserveWhitespaces: false,
                        styles: ["\n    .alpha-picker {\n      position: relative;\n    }\n    .color-alpha {\n      radius: 2px;\n    }\n  "]
                    }] }
        ];
        /** @nocollapse */
        AlphaPickerComponent.ctorParameters = function () { return []; };
        AlphaPickerComponent.propDecorators = {
            width: [{ type: core.Input }],
            height: [{ type: core.Input }],
            direction: [{ type: core.Input }]
        };
        return AlphaPickerComponent;
    }(ngxColor.ColorWrap));
    var ColorAlphaModule = /** @class */ (function () {
        function ColorAlphaModule() {
        }
        ColorAlphaModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [AlphaPickerComponent],
                        exports: [AlphaPickerComponent],
                        imports: [common.CommonModule, ngxColor.AlphaModule, ngxColor.CheckboardModule],
                    },] }
        ];
        return ColorAlphaModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.AlphaPickerComponent = AlphaPickerComponent;
    exports.ColorAlphaModule = ColorAlphaModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
