// ngx-color/photoshop/esm2015/photoshop-button.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
export class PhotoshopButtonComponent {
    constructor() {
        this.label = '';
        this.active = false;
        this.click = new EventEmitter();
    }
}
PhotoshopButtonComponent.decorators = [
    { type: Component, args: [{
                selector: 'color-photoshop-button',
                template: `
    <div class="photoshop-button"  [class.active]="active"
      (click)="click.emit($event)"
    >
      {{ label }}
    </div>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                styles: [`
    .photoshop-button {
      background-image: linear-gradient(
        -180deg,
        rgb(255, 255, 255) 0%,
        rgb(230, 230, 230) 100%
      );
      border: 1px solid rgb(135, 135, 135);
      border-radius: 2px;
      height: 22px;
      box-shadow: rgb(234, 234, 234) 0px 1px 0px 0px;
      font-size: 14px;
      color: rgb(0, 0, 0);
      line-height: 20px;
      text-align: center;
      margin-bottom: 10px;
      cursor: pointer;
    }
    .photoshop-button.active {
      box-shadow: 0 0 0 1px #878787;
    }
  `]
            }] }
];
PhotoshopButtonComponent.propDecorators = {
    label: [{ type: Input }],
    active: [{ type: Input }],
    click: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    PhotoshopButtonComponent.prototype.label;
    /** @type {?} */
    PhotoshopButtonComponent.prototype.active;
    /** @type {?} */
    PhotoshopButtonComponent.prototype.click;
}
