// ngx-color/compact/esm2015/compact-fields.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { isValidHex } from 'ngx-color';
export class CompactFieldsComponent {
    constructor() {
        this.onChange = new EventEmitter();
        this.HEXWrap = {
            marginTop: '-3px',
            marginBottom: '-3px',
            // flex: '6 1 0%',
            position: 'relative',
        };
        this.HEXinput = {
            width: '80%',
            padding: '0px',
            paddingLeft: '20%',
            border: 'none',
            outline: 'none',
            background: 'none',
            fontSize: '12px',
            color: '#333',
            height: '16px',
        };
        this.HEXlabel = {
            display: 'none',
        };
        this.RGBwrap = {
            marginTop: '-3px',
            marginBottom: '-3px',
            // flex: '3 1 0%',
            position: 'relative',
        };
        this.RGBinput = {
            width: '80%',
            padding: '0px',
            paddingLeft: '30%',
            border: 'none',
            outline: 'none',
            background: 'none',
            fontSize: '12px',
            color: '#333',
            height: '16px',
        };
        this.RGBlabel = {
            position: 'absolute',
            top: '6px',
            left: '0px',
            'line-height': '16px',
            'text-transform': 'uppercase',
            fontSize: '12px',
            color: '#999',
        };
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    handleChange({ data, $event }) {
        if (data.hex) {
            if (isValidHex(data.hex)) {
                this.onChange.emit({
                    data: {
                        hex: data.hex,
                        source: 'hex',
                    },
                    $event,
                });
            }
        }
        else {
            this.onChange.emit({
                data: {
                    r: data.r || this.rgb.r,
                    g: data.g || this.rgb.g,
                    b: data.b || this.rgb.b,
                    source: 'rgb',
                },
                $event,
            });
        }
    }
}
CompactFieldsComponent.decorators = [
    { type: Component, args: [{
                selector: 'color-compact-fields',
                template: `
  <div class="compact-fields">
    <div class="compact-active" [style.background]="hex"></div>
    <color-editable-input
      style="flex: 6 1 0%;"
      [style]="{ wrap: HEXWrap, input: HEXinput, label: HEXlabel }"
      label="hex"
      [value]="hex"
      (onChange)="handleChange($event)"
    ></color-editable-input>
    <color-editable-input
      style="flex: 3 1 0%"
      [style]="{ wrap: RGBwrap, input: RGBinput, label: RGBlabel }"
      label="r"
      [value]="rgb.r"
      (onChange)="handleChange($event)"
    ></color-editable-input>
    <color-editable-input
      style="flex: 3 1 0%"
      [style]="{ wrap: RGBwrap, input: RGBinput, label: RGBlabel }"
      label="g"
      [value]="rgb.g"
      (onChange)="handleChange($event)"
    ></color-editable-input>
    <color-editable-input
      style="flex: 3 1 0%"
      [style]="{ wrap: RGBwrap, input: RGBinput, label: RGBlabel }"
      label="b"
      [value]="rgb.b"
      (onChange)="handleChange($event)"
    ></color-editable-input>
  </div>
  `,
                changeDetection: ChangeDetectionStrategy.OnPush,
                preserveWhitespaces: false,
                styles: [`
  .compact-fields {
    display: flex;
    padding-bottom: 6px;
    padding-right: 5px;
    position: relative;
  }
  .compact-active {
    position: absolute;
    top: 6px;
    left: 5px;
    height: 9px;
    width: 9px;
  }
  `]
            }] }
];
CompactFieldsComponent.propDecorators = {
    hex: [{ type: Input }],
    rgb: [{ type: Input }],
    onChange: [{ type: Output }]
};
if (false) {
    /** @type {?} */
    CompactFieldsComponent.prototype.hex;
    /** @type {?} */
    CompactFieldsComponent.prototype.rgb;
    /** @type {?} */
    CompactFieldsComponent.prototype.onChange;
    /** @type {?} */
    CompactFieldsComponent.prototype.HEXWrap;
    /** @type {?} */
    CompactFieldsComponent.prototype.HEXinput;
    /** @type {?} */
    CompactFieldsComponent.prototype.HEXlabel;
    /** @type {?} */
    CompactFieldsComponent.prototype.RGBwrap;
    /** @type {?} */
    CompactFieldsComponent.prototype.RGBinput;
    /** @type {?} */
    CompactFieldsComponent.prototype.RGBlabel;
}
