// ngx-color/fesm5/ngx-color.js
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, Directive, ElementRef, HostListener, Output, EventEmitter } from '@angular/core';
import { Subject, fromEvent } from 'rxjs';
import { distinctUntilChanged, debounceTime } from 'rxjs/operators';
import * as _tinycolor from 'tinycolor2';
import { __assign } from 'tslib';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var checkboardCache = {};
/**
 * @param {?} c1
 * @param {?} c2
 * @param {?} size
 * @return {?}
 */
function render(c1, c2, size) {
    if (typeof document === 'undefined') {
        return null;
    }
    /** @type {?} */
    var canvas = document.createElement('canvas');
    canvas.width = size * 2;
    canvas.height = size * 2;
    /** @type {?} */
    var ctx = canvas.getContext('2d');
    if (!ctx) {
        return null;
    } // If no context can be found, return early.
    ctx.fillStyle = c1;
    ctx.fillRect(0, 0, canvas.width, canvas.height);
    ctx.fillStyle = c2;
    ctx.fillRect(0, 0, size, size);
    ctx.translate(size, size);
    ctx.fillRect(0, 0, size, size);
    return canvas.toDataURL();
}
/**
 * @param {?} c1
 * @param {?} c2
 * @param {?} size
 * @return {?}
 */
function getCheckerboard(c1, c2, size) {
    /** @type {?} */
    var key = c1 + "-" + c2 + "-" + size;
    if (checkboardCache[key]) {
        return checkboardCache[key];
    }
    /** @type {?} */
    var checkboard = render(c1, c2, size);
    if (!checkboard) {
        return null;
    }
    checkboardCache[key] = checkboard;
    return checkboard;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var CheckboardComponent = /** @class */ (function () {
    function CheckboardComponent() {
        this.white = 'transparent';
        this.size = 8;
        this.grey = 'rgba(0,0,0,.08)';
    }
    /**
     * @return {?}
     */
    CheckboardComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        /** @type {?} */
        var background = getCheckerboard(this.white, this.grey, this.size);
        this.gridStyles = {
            borderRadius: this.borderRadius,
            boxShadow: this.boxShadow,
            background: "url(" + background + ") center left",
        };
    };
    CheckboardComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-checkboard',
                    template: "<div class=\"grid\" [ngStyle]=\"gridStyles\"></div>",
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n  .grid {\n    top: 0px;\n    right: 0px;\n    bottom: 0px;\n    left: 0px;\n    position: absolute;\n  }\n  "]
                }] }
    ];
    CheckboardComponent.propDecorators = {
        white: [{ type: Input }],
        size: [{ type: Input }],
        grey: [{ type: Input }],
        boxShadow: [{ type: Input }],
        borderRadius: [{ type: Input }]
    };
    return CheckboardComponent;
}());
var CheckboardModule = /** @class */ (function () {
    function CheckboardModule() {
    }
    CheckboardModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [CheckboardComponent],
                    exports: [CheckboardComponent],
                    imports: [CommonModule],
                },] }
    ];
    return CheckboardModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var CoordinatesDirective = /** @class */ (function () {
    function CoordinatesDirective(el) {
        this.el = el;
        this.coordinatesChange = new Subject();
        this.mousechange = new Subject();
        this.mouseListening = false;
    }
    /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    CoordinatesDirective.prototype.mousemove = /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    function ($event, x, y, isTouch) {
        if (isTouch === void 0) { isTouch = false; }
        if (this.mouseListening) {
            $event.preventDefault();
            this.mousechange.next({ $event: $event, x: x, y: y, isTouch: isTouch });
        }
    };
    /**
     * @return {?}
     */
    CoordinatesDirective.prototype.mouseup = /**
     * @return {?}
     */
    function () {
        this.mouseListening = false;
    };
    /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    CoordinatesDirective.prototype.mousedown = /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    function ($event, x, y, isTouch) {
        if (isTouch === void 0) { isTouch = false; }
        $event.preventDefault();
        this.mouseListening = true;
        this.mousechange.next({ $event: $event, x: x, y: y, isTouch: isTouch });
    };
    /**
     * @return {?}
     */
    CoordinatesDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.sub = this.mousechange
            .pipe(
        // limit times it is updated for the same area
        distinctUntilChanged(function (p, q) { return p.x === q.x && p.y === q.y; }))
            .subscribe(function (n) { return _this.handleChange(n.x, n.y, n.$event, n.isTouch); });
    };
    /**
     * @return {?}
     */
    CoordinatesDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.sub.unsubscribe();
    };
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} $event
     * @param {?} isTouch
     * @return {?}
     */
    CoordinatesDirective.prototype.handleChange = /**
     * @param {?} x
     * @param {?} y
     * @param {?} $event
     * @param {?} isTouch
     * @return {?}
     */
    function (x, y, $event, isTouch) {
        /** @type {?} */
        var containerWidth = this.el.nativeElement.clientWidth;
        /** @type {?} */
        var containerHeight = this.el.nativeElement.clientHeight;
        /** @type {?} */
        var left = x -
            (this.el.nativeElement.getBoundingClientRect().left + window.pageXOffset);
        /** @type {?} */
        var top = y - this.el.nativeElement.getBoundingClientRect().top;
        if (!isTouch) {
            top = top - window.pageYOffset;
        }
        this.coordinatesChange.next({
            x: x,
            y: y,
            top: top,
            left: left,
            containerWidth: containerWidth,
            containerHeight: containerHeight,
            $event: $event,
        });
    };
    CoordinatesDirective.decorators = [
        { type: Directive, args: [{ selector: '[ngx-color-coordinates]' },] }
    ];
    /** @nocollapse */
    CoordinatesDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    CoordinatesDirective.propDecorators = {
        coordinatesChange: [{ type: Output }],
        mousemove: [{ type: HostListener, args: ['window:mousemove', ['$event', '$event.pageX', '$event.pageY'],] }, { type: HostListener, args: ['window:touchmove', [
                        '$event',
                        '$event.touches[0].clientX',
                        '$event.touches[0].clientY',
                        'true',
                    ],] }],
        mouseup: [{ type: HostListener, args: ['window:mouseup',] }, { type: HostListener, args: ['window:touchend',] }],
        mousedown: [{ type: HostListener, args: ['mousedown', ['$event', '$event.pageX', '$event.pageY'],] }, { type: HostListener, args: ['touchstart', [
                        '$event',
                        '$event.touches[0].clientX',
                        '$event.touches[0].clientY',
                        'true',
                    ],] }]
    };
    return CoordinatesDirective;
}());
var CoordinatesModule = /** @class */ (function () {
    function CoordinatesModule() {
    }
    CoordinatesModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [CoordinatesDirective],
                    exports: [CoordinatesDirective],
                },] }
    ];
    return CoordinatesModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var AlphaComponent = /** @class */ (function () {
    function AlphaComponent() {
        this.direction = 'horizontal';
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    AlphaComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (this.direction === 'vertical') {
            this.pointerLeft = 0;
            this.pointerTop = this.rgb.a * 100;
            this.gradient = {
                background: "linear-gradient(to bottom, rgba(" + this.rgb.r + "," + this.rgb.g + "," + this.rgb.b + ", 0) 0%,\n          rgba(" + this.rgb.r + "," + this.rgb.g + "," + this.rgb.b + ", 1) 100%)",
            };
        }
        else {
            this.gradient = {
                background: "linear-gradient(to right, rgba(" + this.rgb.r + "," + this.rgb.g + "," + this.rgb.b + ", 0) 0%,\n          rgba(" + this.rgb.r + "," + this.rgb.g + "," + this.rgb.b + ", 1) 100%)",
            };
            this.pointerLeft = this.rgb.a * 100;
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    AlphaComponent.prototype.handleChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var top = _a.top, left = _a.left, containerHeight = _a.containerHeight, containerWidth = _a.containerWidth, $event = _a.$event;
        /** @type {?} */
        var data;
        if (this.direction === 'vertical') {
            /** @type {?} */
            var a = void 0;
            if (top < 0) {
                a = 0;
            }
            else if (top > containerHeight) {
                a = 1;
            }
            else {
                a = Math.round(top * 100 / containerHeight) / 100;
            }
            if (this.hsl.a !== a) {
                data = {
                    h: this.hsl.h,
                    s: this.hsl.s,
                    l: this.hsl.l,
                    a: a,
                    source: 'rgb',
                };
            }
        }
        else {
            /** @type {?} */
            var a = void 0;
            if (left < 0) {
                a = 0;
            }
            else if (left > containerWidth) {
                a = 1;
            }
            else {
                a = Math.round(left * 100 / containerWidth) / 100;
            }
            if (this.hsl.a !== a) {
                data = {
                    h: this.hsl.h,
                    s: this.hsl.s,
                    l: this.hsl.l,
                    a: a,
                    source: 'rgb',
                };
            }
        }
        if (!data) {
            return null;
        }
        this.onChange.emit({ data: data, $event: $event });
    };
    AlphaComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-alpha',
                    template: "\n  <div class=\"alpha\" [style.border-radius]=\"radius\">\n    <div class=\"alpha-checkboard\">\n      <color-checkboard></color-checkboard>\n    </div>\n    <div class=\"alpha-gradient\" [ngStyle]=\"gradient\" [style.box-shadow]=\"shadow\" [style.border-radius]=\"radius\"></div>\n    <div ngx-color-coordinates (coordinatesChange)=\"handleChange($event)\" class=\"alpha-container color-alpha-{{direction}}\">\n      <div class=\"alpha-pointer\" [style.left.%]=\"pointerLeft\" [style.top.%]=\"pointerTop\">\n        <div class=\"alpha-slider\" [ngStyle]=\"pointer\"></div>\n      </div>\n    </div>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n    .alpha {\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .alpha-checkboard {\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n      overflow: hidden;\n    }\n    .alpha-gradient {\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .alpha-container {\n      position: relative;\n      height: 100%;\n      margin: 0 3px;\n    }\n    .alpha-pointer {\n      position: absolute;\n    }\n    .alpha-slider {\n      width: 4px;\n      border-radius: 1px;\n      height: 8px;\n      box-shadow: 0 0 2px rgba(0, 0, 0, .6);\n      background: #fff;\n      margin-top: 1px;\n      transform: translateX(-2px);\n    },\n  "]
                }] }
    ];
    AlphaComponent.propDecorators = {
        hsl: [{ type: Input }],
        rgb: [{ type: Input }],
        pointer: [{ type: Input }],
        shadow: [{ type: Input }],
        radius: [{ type: Input }],
        direction: [{ type: Input }],
        onChange: [{ type: Output }]
    };
    return AlphaComponent;
}());
var AlphaModule = /** @class */ (function () {
    function AlphaModule() {
    }
    AlphaModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [AlphaComponent],
                    exports: [AlphaComponent],
                    imports: [CommonModule, CheckboardModule, CoordinatesModule],
                },] }
    ];
    return AlphaModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/** @type {?} */
var tinycolor = _tinycolor;
/**
 * @param {?} data
 * @return {?}
 */
function simpleCheckForValidColor(data) {
    /** @type {?} */
    var keysToCheck = ['r', 'g', 'b', 'a', 'h', 's', 'l', 'v'];
    /** @type {?} */
    var checked = 0;
    /** @type {?} */
    var passed = 0;
    keysToCheck.forEach(function (letter) {
        if (!data[letter]) {
            return;
        }
        checked += 1;
        if (!isNaN(data[letter])) {
            passed += 1;
        }
        if (letter === 's' || letter === 'l') {
            /** @type {?} */
            var percentPatt = /^\d+%$/;
            if (percentPatt.test(data[letter])) {
                passed += 1;
            }
        }
    });
    return checked === passed ? data : false;
}
/**
 * @param {?} data
 * @param {?=} oldHue
 * @return {?}
 */
function toState(data, oldHue) {
    /** @type {?} */
    var color = data.hex ? tinycolor(data.hex) : tinycolor(data);
    /** @type {?} */
    var hsl = color.toHsl();
    /** @type {?} */
    var hsv = color.toHsv();
    /** @type {?} */
    var rgb = color.toRgb();
    /** @type {?} */
    var hex = color.toHex();
    if (hsl.s === 0) {
        hsl.h = oldHue || 0;
        hsv.h = oldHue || 0;
    }
    /** @type {?} */
    var transparent = hex === '000000' && rgb.a === 0;
    return {
        hsl: hsl,
        hex: transparent ? 'transparent' : "#" + hex,
        rgb: rgb,
        hsv: hsv,
        oldHue: data.h || oldHue || hsl.h,
        source: data.source,
    };
}
/**
 * @param {?} hex
 * @return {?}
 */
function isValidHex(hex) {
    return tinycolor(hex).isValid();
}
/**
 * @param {?} data
 * @return {?}
 */
function getContrastingColor(data) {
    if (!data) {
        return '#fff';
    }
    /** @type {?} */
    var col = toState(data);
    if (col.hex === 'transparent') {
        return 'rgba(0,0,0,0.4)';
    }
    /** @type {?} */
    var yiq = (col.rgb.r * 299 + col.rgb.g * 587 + col.rgb.b * 114) / 1000;
    return yiq >= 128 ? '#000' : '#fff';
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var ColorWrap = /** @class */ (function () {
    function ColorWrap() {
        this.className = '';
        this.color = {
            h: 250,
            s: 0.5,
            l: 0.2,
            a: 1,
        };
        this.onChange = new EventEmitter();
        this.onChangeComplete = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ColorWrap.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.changes = this.onChange.pipe(debounceTime(100), distinctUntilChanged())
            .subscribe(function (x) { return _this.onChangeComplete.emit(x); });
        this.setState(toState(this.color, 0));
        this.currentColor = this.hex;
    };
    /**
     * @return {?}
     */
    ColorWrap.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.setState(toState(this.color, this.oldHue));
    };
    /**
     * @return {?}
     */
    ColorWrap.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.changes.unsubscribe();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ColorWrap.prototype.setState = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.oldHue = data.oldHue;
        this.hsl = data.hsl;
        this.hsv = data.hsv;
        this.rgb = data.rgb;
        this.hex = data.hex;
        this.source = data.source;
        this.afterValidChange();
    };
    /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    ColorWrap.prototype.handleChange = /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    function (data, $event) {
        /** @type {?} */
        var isValidColor = simpleCheckForValidColor(data);
        if (isValidColor) {
            /** @type {?} */
            var color = toState(data, data.h || this.oldHue);
            this.setState(color);
            this.onChange.emit({ color: color, $event: $event });
            this.afterValidChange();
        }
    };
    /** hook for components after a complete change */
    /**
     * hook for components after a complete change
     * @return {?}
     */
    ColorWrap.prototype.afterValidChange = /**
     * hook for components after a complete change
     * @return {?}
     */
    function () { };
    /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    ColorWrap.prototype.handleSwatchHover = /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    function (data, $event) {
        /** @type {?} */
        var isValidColor = simpleCheckForValidColor(data);
        if (isValidColor) {
            /** @type {?} */
            var color = toState(data, data.h || this.oldHue);
            this.setState(color);
            this.onSwatchHover.emit({ color: color, $event: $event });
        }
    };
    ColorWrap.decorators = [
        { type: Component, args: [{
                    // create seletor base for test override property
                    selector: 'color-wrap',
                    template: ""
                }] }
    ];
    ColorWrap.propDecorators = {
        className: [{ type: Input }],
        color: [{ type: Input }],
        onChange: [{ type: Output }],
        onChangeComplete: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return ColorWrap;
}());
var ColorWrapModule = /** @class */ (function () {
    function ColorWrapModule() {
    }
    ColorWrapModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ColorWrap],
                    exports: [ColorWrap],
                    imports: [CommonModule],
                },] }
    ];
    return ColorWrapModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var EditableInputComponent = /** @class */ (function () {
    function EditableInputComponent() {
        this.placeholder = '';
        this.onChange = new EventEmitter();
        this.focus = false;
    }
    /**
     * @return {?}
     */
    EditableInputComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.wrapStyle = this.style && this.style.wrap ? this.style.wrap : {};
        this.inputStyle = this.style && this.style.input ? this.style.input : {};
        this.labelStyle = this.style && this.style.label ? this.style.label : {};
        if (this.dragLabel) {
            this.labelStyle.cursor = 'ew-resize';
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EditableInputComponent.prototype.handleFocus = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.focus = true;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EditableInputComponent.prototype.handleFocusOut = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.focus = false;
        this.currentValue = this.blurValue;
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EditableInputComponent.prototype.handleKeydown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _a, _b;
        // In case `e.target.value` is a percentage remove the `%` character
        // and update accordingly with a percentage
        // https://github.com/casesandberg/react-color/issues/383
        /** @type {?} */
        var stringValue = String($event.target.value);
        /** @type {?} */
        var isPercentage = stringValue.indexOf('%') > -1;
        /** @type {?} */
        var number = Number(stringValue.replace(/%/g, ''));
        if (!isNaN(number)) {
            /** @type {?} */
            var amount = this.arrowOffset || 1;
            // Up
            if ($event.keyCode === 38) {
                if (this.label) {
                    this.onChange.emit({
                        data: (_a = {}, _a[this.label] = number + amount, _a),
                        $event: $event,
                    });
                }
                else {
                    this.onChange.emit({ data: number + amount, $event: $event });
                }
                if (isPercentage) {
                    this.currentValue = number + amount + "%";
                }
                else {
                    this.currentValue = number + amount;
                }
            }
            // Down
            if ($event.keyCode === 40) {
                if (this.label) {
                    this.onChange.emit({
                        data: (_b = {}, _b[this.label] = number - amount, _b),
                        $event: $event,
                    });
                }
                else {
                    this.onChange.emit({ data: number - amount, $event: $event });
                }
                if (isPercentage) {
                    this.currentValue = number - amount + "%";
                }
                else {
                    this.currentValue = number - amount;
                }
            }
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EditableInputComponent.prototype.handleKeyup = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _a;
        if ($event.keyCode === 40 || $event.keyCode === 38) {
            return;
        }
        if (this.label) {
            this.onChange.emit({
                data: (_a = {}, _a[this.label] = $event.target.value, _a),
                $event: $event,
            });
        }
        else {
            this.onChange.emit({ data: $event.target.value, $event: $event });
        }
    };
    /**
     * @return {?}
     */
    EditableInputComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (!this.focus) {
            this.currentValue = String(this.value).toUpperCase();
            this.blurValue = String(this.value).toUpperCase();
        }
        else {
            this.blurValue = String(this.value).toUpperCase();
        }
    };
    /**
     * @return {?}
     */
    EditableInputComponent.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.unsubscribe();
    };
    /**
     * @return {?}
     */
    EditableInputComponent.prototype.subscribe = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.mousemove = fromEvent(document, 'mousemove').subscribe(function (ev) {
            return _this.handleDrag(ev);
        });
        this.mouseup = fromEvent(document, 'mouseup').subscribe(function () {
            return _this.unsubscribe();
        });
    };
    /**
     * @return {?}
     */
    EditableInputComponent.prototype.unsubscribe = /**
     * @return {?}
     */
    function () {
        if (this.mousemove) {
            this.mousemove.unsubscribe();
        }
        if (this.mouseup) {
            this.mouseup.unsubscribe();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EditableInputComponent.prototype.handleMousedown = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        if (this.dragLabel) {
            $event.preventDefault();
            this.handleDrag($event);
            this.subscribe();
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    EditableInputComponent.prototype.handleDrag = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        var _a;
        if (this.dragLabel) {
            /** @type {?} */
            var newValue = Math.round(this.value + $event.movementX);
            if (newValue >= 0 && newValue <= this.dragMax) {
                this.onChange.emit({ data: (_a = {}, _a[this.label] = newValue, _a), $event: $event });
            }
        }
    };
    EditableInputComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-editable-input',
                    template: "\n  <div class=\"wrap\" [ngStyle]=\"wrapStyle\">\n    <input [ngStyle]=\"inputStyle\" spellCheck=\"false\"\n      [value]=\"currentValue\" [placeholder]=\"placeholder\"\n      (keydown)=\"handleKeydown($event)\" (keyup)=\"handleKeyup($event)\"\n      (focus)=\"handleFocus($event)\" (focusout)=\"handleFocusOut($event)\" />\n    <span *ngIf=\"label\" [ngStyle]=\"labelStyle\" (mousedown)=\"handleMousedown($event)\">\n      {{ label }}\n    </span>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n    :host {\n      display: flex;\n    }\n    .wrap {\n      position: relative;\n    }\n  "]
                }] }
    ];
    EditableInputComponent.propDecorators = {
        style: [{ type: Input }],
        label: [{ type: Input }],
        value: [{ type: Input }],
        arrowOffset: [{ type: Input }],
        dragLabel: [{ type: Input }],
        dragMax: [{ type: Input }],
        placeholder: [{ type: Input }],
        onChange: [{ type: Output }]
    };
    return EditableInputComponent;
}());
var EditableInputModule = /** @class */ (function () {
    function EditableInputModule() {
    }
    EditableInputModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [EditableInputComponent],
                    exports: [EditableInputComponent],
                    imports: [CommonModule],
                },] }
    ];
    return EditableInputModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var HueComponent = /** @class */ (function () {
    function HueComponent() {
        this.hidePointer = false;
        this.direction = 'horizontal';
        this.onChange = new EventEmitter();
        this.left = '0px';
        this.top = '';
    }
    /**
     * @return {?}
     */
    HueComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (this.direction === 'horizontal') {
            this.left = this.hsl.h * 100 / 360 + "%";
        }
        else {
            this.top = -(this.hsl.h * 100 / 360) + 100 + "%";
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    HueComponent.prototype.handleChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var top = _a.top, left = _a.left, containerHeight = _a.containerHeight, containerWidth = _a.containerWidth, $event = _a.$event;
        /** @type {?} */
        var data;
        if (this.direction === 'vertical') {
            /** @type {?} */
            var h = void 0;
            if (top < 0) {
                h = 359;
            }
            else if (top > containerHeight) {
                h = 0;
            }
            else {
                /** @type {?} */
                var percent = -(top * 100 / containerHeight) + 100;
                h = 360 * percent / 100;
            }
            if (this.hsl.h !== h) {
                data = {
                    h: h,
                    s: this.hsl.s,
                    l: this.hsl.l,
                    a: this.hsl.a,
                    source: 'rgb',
                };
            }
        }
        else {
            /** @type {?} */
            var h = void 0;
            if (left < 0) {
                h = 0;
            }
            else if (left > containerWidth) {
                h = 359;
            }
            else {
                /** @type {?} */
                var percent = left * 100 / containerWidth;
                h = 360 * percent / 100;
            }
            if (this.hsl.h !== h) {
                data = {
                    h: h,
                    s: this.hsl.s,
                    l: this.hsl.l,
                    a: this.hsl.a,
                    source: 'rgb',
                };
            }
        }
        if (!data) {
            return null;
        }
        this.onChange.emit({ data: data, $event: $event });
    };
    HueComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-hue',
                    template: "\n  <div class=\"color-hue color-hue-{{direction}}\" [style.border-radius.px]=\"radius\" [style.box-shadow]=\"shadow\">\n    <div ngx-color-coordinates (coordinatesChange)=\"handleChange($event)\" class=\"color-hue-container\">\n      <div class=\"color-hue-pointer\" [style.left]=\"left\" [style.top]=\"top\" *ngIf=\"!hidePointer\">\n        <div class=\"color-hue-slider\" [ngStyle]=\"pointer\"></div>\n      </div>\n    </div>\n  </div>\n  ",
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n    .color-hue {\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .color-hue-container {\n      margin: 0 2px;\n      position: relative;\n      height: 100%;\n    }\n    .color-hue-pointer {\n      position: absolute;\n    }\n    .color-hue-slider {\n      margin-top: 1px;\n      width: 4px;\n      border-radius: 1px;\n      height: 8px;\n      box-shadow: 0 0 2px rgba(0, 0, 0, .6);\n      background: #fff;\n      transform: translateX(-2px);\n    }\n    .color-hue-horizontal {\n      background: linear-gradient(to right, #f00 0%, #ff0 17%, #0f0\n        33%, #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);\n    }\n    .color-hue-vertical {\n      background: linear-gradient(to top, #f00 0%, #ff0 17%, #0f0 33%,\n        #0ff 50%, #00f 67%, #f0f 83%, #f00 100%);\n    }\n  "]
                }] }
    ];
    HueComponent.propDecorators = {
        hsl: [{ type: Input }],
        pointer: [{ type: Input }],
        radius: [{ type: Input }],
        shadow: [{ type: Input }],
        hidePointer: [{ type: Input }],
        direction: [{ type: Input }],
        onChange: [{ type: Output }]
    };
    return HueComponent;
}());
var HueModule = /** @class */ (function () {
    function HueModule() {
    }
    HueModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [HueComponent],
                    exports: [HueComponent],
                    imports: [CommonModule, CoordinatesModule],
                },] }
    ];
    return HueModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var RaisedComponent = /** @class */ (function () {
    function RaisedComponent() {
        this.zDepth = 1;
        this.radius = 1;
        this.background = '#fff';
    }
    RaisedComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-raised',
                    template: "\n  <div class=\"raised-wrap\">\n    <div class=\"raised-bg zDepth-{{zDepth}}\" [style.background]=\"background\"></div>\n    <div class=\"raised-content\">\n      <ng-content></ng-content>\n    </div>\n  </div>\n  ",
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n    .raised-wrap {\n      position: relative;\n      display: inline-block;\n    }\n    .raised-bg {\n      position: absolute;\n      top: 0px;\n      right: 0px;\n      bottom: 0px;\n      left: 0px;\n    }\n    .raised-content {\n      position: relative;\n    }\n    .zDepth-0 {\n      box-shadow: none;\n    }\n    .zDepth-1 {\n      box-shadow: 0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16);\n    }\n    .zDepth-2 {\n      box-shadow: 0 6px 20px rgba(0,0,0,.19), 0 8px 17px rgba(0,0,0,.2);\n    }\n    .zDepth-3 {\n      box-shadow: 0 17px 50px rgba(0,0,0,.19), 0 12px 15px rgba(0,0,0,.24);\n    }\n    .zDepth-4 {\n      box-shadow: 0 25px 55px rgba(0,0,0,.21), 0 16px 28px rgba(0,0,0,.22);\n    }\n    .zDepth-5 {\n      box-shadow: 0 40px 77px rgba(0,0,0,.22), 0 27px 24px rgba(0,0,0,.2);\n    }\n  "]
                }] }
    ];
    RaisedComponent.propDecorators = {
        zDepth: [{ type: Input }],
        radius: [{ type: Input }],
        background: [{ type: Input }]
    };
    return RaisedComponent;
}());
var RaisedModule = /** @class */ (function () {
    function RaisedModule() {
    }
    RaisedModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [RaisedComponent],
                    exports: [RaisedComponent],
                    imports: [CommonModule],
                },] }
    ];
    return RaisedModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var SaturationComponent = /** @class */ (function () {
    function SaturationComponent() {
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SaturationComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.background = "hsl(" + this.hsl.h + ", 100%, 50%)";
        this.pointerTop = -(this.hsv.v * 100) + 1 + 100 + '%';
        this.pointerLeft = this.hsv.s * 100 + '%';
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    SaturationComponent.prototype.handleChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var top = _a.top, left = _a.left, containerHeight = _a.containerHeight, containerWidth = _a.containerWidth, $event = _a.$event;
        if (left < 0) {
            left = 0;
        }
        else if (left > containerWidth) {
            left = containerWidth;
        }
        else if (top < 0) {
            top = 0;
        }
        else if (top > containerHeight) {
            top = containerHeight;
        }
        /** @type {?} */
        var saturation = left / containerWidth;
        /** @type {?} */
        var bright = -(top / containerHeight) + 1;
        bright = bright > 0 ? bright : 0;
        bright = bright > 1 ? 1 : bright;
        /** @type {?} */
        var data = {
            h: this.hsl.h,
            s: saturation,
            v: bright,
            a: this.hsl.a,
            source: 'hsva',
        };
        this.onChange.emit({ data: data, $event: $event });
    };
    SaturationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-saturation',
                    template: "\n  <div class=\"color-saturation\" ngx-color-coordinates (coordinatesChange)=\"handleChange($event)\" [style.background]=\"background\">\n    <div class=\"saturation-white\">\n      <div class=\"saturation-black\"></div>\n      <div class=\"saturation-pointer\" [ngStyle]=\"pointer\" [style.top]=\"pointerTop\" [style.left]=\"pointerLeft\">\n        <div class=\"saturation-circle\" [ngStyle]=\"circle\"></div>\n      </div>\n    </div>\n  </div>\n  ",
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n    .saturation-white {\n      background: linear-gradient(to right, #fff, rgba(255,255,255,0));\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .saturation-black {\n      background: linear-gradient(to top, #000, rgba(0,0,0,0));\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .color-saturation {\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .saturation-pointer {\n      position: absolute;\n      cursor: default;\n    }\n    .saturation-circle {\n      width: 4px;\n      height: 4px;\n      box-shadow: 0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3), 0 0 1px 2px rgba(0,0,0,.4);\n      border-radius: 50%;\n      cursor: hand;\n      transform: translate(-2px, -4px);\n    }\n  "]
                }] }
    ];
    SaturationComponent.propDecorators = {
        hsl: [{ type: Input }],
        hsv: [{ type: Input }],
        radius: [{ type: Input }],
        pointer: [{ type: Input }],
        circle: [{ type: Input }],
        onChange: [{ type: Output }]
    };
    return SaturationComponent;
}());
var SaturationModule = /** @class */ (function () {
    function SaturationModule() {
    }
    SaturationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [SaturationComponent],
                    exports: [SaturationComponent],
                    imports: [CommonModule, CoordinatesModule],
                },] }
    ];
    return SaturationModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
var SwatchComponent = /** @class */ (function () {
    function SwatchComponent() {
        this.style = {};
        this.focusStyle = {};
        this.onClick = new EventEmitter();
        this.onHover = new EventEmitter();
        this.divStyles = {};
        this.focusStyles = {};
        this.inFocus = false;
    }
    /**
     * @return {?}
     */
    SwatchComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.divStyles = __assign({ background: this.color, height: '100%', width: '100%', cursor: 'pointer', position: 'relative', outline: 'none' }, this.style);
        this.focusStyles = __assign({}, this.divStyles, this.focusStyle);
    };
    /**
     * @return {?}
     */
    SwatchComponent.prototype.activeStyles = /**
     * @return {?}
     */
    function () {
        return this.focus || this.inFocus ? this.focusStyles : this.divStyles;
    };
    /**
     * @return {?}
     */
    SwatchComponent.prototype.handleFocusOut = /**
     * @return {?}
     */
    function () {
        this.inFocus = false;
    };
    /**
     * @return {?}
     */
    SwatchComponent.prototype.handleFocus = /**
     * @return {?}
     */
    function () {
        this.inFocus = true;
    };
    /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    SwatchComponent.prototype.handleHover = /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    function (hex, $event) {
        this.onHover.emit({ hex: hex, $event: $event });
    };
    /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    SwatchComponent.prototype.handleClick = /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    function (hex, $event) {
        this.onClick.emit({ hex: hex, $event: $event });
    };
    SwatchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-swatch',
                    template: "\n  <div class=\"swatch\"\n    [ngStyle]=\"activeStyles()\" [attr.title]=\"color\"\n    (click)=\"handleClick(color, $event)\"\n    (keydown.enter)=\"handleClick(color, $event)\"\n    (focus)=\"handleFocus()\"\n    (focusout)=\"handleFocusOut()\"\n    (mouseover)=\"handleHover(color, $event)\"\n    tabindex=\"0\"\n  >\n    <ng-content></ng-content>\n    <color-checkboard\n      *ngIf=\"color === 'transparent'\"\n      boxShadow=\"inset 0 0 0 1px rgba(0,0,0,0.1)\"\n    ></color-checkboard>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false
                }] }
    ];
    SwatchComponent.propDecorators = {
        color: [{ type: Input }],
        style: [{ type: Input }],
        focusStyle: [{ type: Input }],
        focus: [{ type: Input }],
        onClick: [{ type: Output }],
        onHover: [{ type: Output }]
    };
    return SwatchComponent;
}());
var SwatchModule = /** @class */ (function () {
    function SwatchModule() {
    }
    SwatchModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [SwatchComponent],
                    exports: [SwatchComponent],
                    imports: [CommonModule, CheckboardModule],
                },] }
    ];
    return SwatchModule;
}());

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */

export { AlphaComponent, AlphaModule, CheckboardComponent, CheckboardModule, ColorWrap, ColorWrapModule, EditableInputComponent, EditableInputModule, HueComponent, HueModule, RaisedComponent, RaisedModule, SaturationComponent, SaturationModule, SwatchComponent, SwatchModule, CoordinatesDirective, CoordinatesModule, render, getCheckerboard, simpleCheckForValidColor, toState, isValidHex, getContrastingColor };
