// ngx-color/circle/esm5/circle-swatch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
var CircleSwatchComponent = /** @class */ (function () {
    function CircleSwatchComponent() {
        this.circleSize = 28;
        this.circleSpacing = 14;
        this.focus = false;
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
        this.swatchStyle = {
            borderRadius: '50%',
            background: 'transparent',
            transition: '100ms box-shadow ease',
        };
    }
    /**
     * @return {?}
     */
    CircleSwatchComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.focusStyle = {
            boxShadow: this.color + " 0px 0px 0px 3px inset",
        };
        this.swatchStyle.boxShadow = "inset 0 0 0 " + this.circleSize / 2 + "px " + this.color;
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    CircleSwatchComponent.prototype.handleClick = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var hex = _a.hex, $event = _a.$event;
        this.onClick.emit({ hex: hex, $event: $event });
    };
    CircleSwatchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-circle-swatch',
                    template: "\n  <div class=\"circle-swatch\"\n    [style.width.px]=\"circleSize\" [style.height.px]=\"circleSize\"\n    [style.margin-right.px]=\"circleSpacing\" [style.margin-bottom.px]=\"circleSpacing\"\n    >\n    <color-swatch\n      [color]=\"color\" [style]=\"swatchStyle\" [focus]=\"focus\" [focusStyle]=\"focusStyle\"\n      (onClick)=\"handleClick($event)\" (onHover)=\"onSwatchHover.emit($event)\">\n    </color-swatch>\n    <div class=\"clear\"></div>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n  .circle-swatch {\n    transform: scale(1);\n    transition: transform 100ms ease;\n  }\n  .circle-swatch:hover {\n    transform: scale(1.2);\n  }\n  "]
                }] }
    ];
    CircleSwatchComponent.propDecorators = {
        color: [{ type: Input }],
        circleSize: [{ type: Input }],
        circleSpacing: [{ type: Input }],
        focus: [{ type: Input }],
        onClick: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return CircleSwatchComponent;
}());
export { CircleSwatchComponent };
if (false) {
    /** @type {?} */
    CircleSwatchComponent.prototype.color;
    /** @type {?} */
    CircleSwatchComponent.prototype.circleSize;
    /** @type {?} */
    CircleSwatchComponent.prototype.circleSpacing;
    /** @type {?} */
    CircleSwatchComponent.prototype.focus;
    /** @type {?} */
    CircleSwatchComponent.prototype.onClick;
    /** @type {?} */
    CircleSwatchComponent.prototype.onSwatchHover;
    /** @type {?} */
    CircleSwatchComponent.prototype.focusStyle;
    /** @type {?} */
    CircleSwatchComponent.prototype.swatchStyle;
}
