// ngx-color/circle/esm5/circle.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { amber, blue, blueGrey, brown, cyan, deepOrange, deepPurple, green, indigo, lightBlue, lightGreen, lime, orange, pink, purple, red, teal, yellow, } from 'material-colors';
import { ColorWrap, SwatchModule } from 'ngx-color';
import { isValidHex } from 'ngx-color';
import { CircleSwatchComponent } from './circle-swatch.component';
var CircleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(CircleComponent, _super);
    function CircleComponent() {
        var _this = _super.call(this) || this;
        /**
         * Pixel value for picker width
         */
        _this.width = 252;
        /**
         * Color squares to display
         */
        _this.colors = [
            red['500'],
            pink['500'],
            purple['500'],
            deepPurple['500'],
            indigo['500'],
            blue['500'],
            lightBlue['500'],
            cyan['500'],
            teal['500'],
            green['500'],
            lightGreen['500'],
            lime['500'],
            yellow['500'],
            amber['500'],
            orange['500'],
            deepOrange['500'],
            brown['500'],
            blueGrey['500'],
        ];
        /**
         * Value for circle size
         */
        _this.circleSize = 28;
        /**
         * Value for spacing between circles
         */
        _this.circleSpacing = 14;
        return _this;
    }
    /**
     * @param {?} color
     * @return {?}
     */
    CircleComponent.prototype.isActive = /**
     * @param {?} color
     * @return {?}
     */
    function (color) {
        return this.hex === color;
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    CircleComponent.prototype.handleBlockChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var hex = _a.hex, $event = _a.$event;
        if (isValidHex(hex)) {
            this.handleChange({ hex: hex, source: 'hex' }, $event);
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    CircleComponent.prototype.handleValueChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var data = _a.data, $event = _a.$event;
        this.handleChange(data, $event);
    };
    CircleComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-circle',
                    template: "\n  <div class=\"circle-picker {{ className }}\"\n    [style.width.px]=\"width\"\n    [style.margin-right.px]=\"-circleSpacing\"\n    [style.margin-bottom.px]=\"-circleSpacing\">\n    <color-circle-swatch\n      *ngFor=\"let color of colors\"\n      [color]=\"color\" [focus]=\"isActive(color)\"\n      (onClick)=\"handleBlockChange($event)\"\n      (onSwatchHover)=\"onSwatchHover.emit($event)\"\n    ></color-circle-swatch>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n      .circle-picker {\n        display: flex;\n        flex-wrap: wrap;\n      }\n    "]
                }] }
    ];
    /** @nocollapse */
    CircleComponent.ctorParameters = function () { return []; };
    CircleComponent.propDecorators = {
        width: [{ type: Input }],
        colors: [{ type: Input }],
        circleSize: [{ type: Input }],
        circleSpacing: [{ type: Input }]
    };
    return CircleComponent;
}(ColorWrap));
export { CircleComponent };
if (false) {
    /**
     * Pixel value for picker width
     * @type {?}
     */
    CircleComponent.prototype.width;
    /**
     * Color squares to display
     * @type {?}
     */
    CircleComponent.prototype.colors;
    /**
     * Value for circle size
     * @type {?}
     */
    CircleComponent.prototype.circleSize;
    /**
     * Value for spacing between circles
     * @type {?}
     */
    CircleComponent.prototype.circleSpacing;
}
var ColorCircleModule = /** @class */ (function () {
    function ColorCircleModule() {
    }
    ColorCircleModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [CircleComponent, CircleSwatchComponent],
                    exports: [CircleComponent, CircleSwatchComponent],
                    imports: [CommonModule, SwatchModule],
                },] }
    ];
    return ColorCircleModule;
}());
export { ColorCircleModule };
