// ngx-color/esm2015/helpers/color.interfaces.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function RGB() { }
if (false) {
    /** @type {?} */
    RGB.prototype.r;
    /** @type {?} */
    RGB.prototype.g;
    /** @type {?} */
    RGB.prototype.b;
}
/**
 * @record
 */
export function RGBA() { }
if (false) {
    /** @type {?} */
    RGBA.prototype.a;
}
/**
 * @record
 */
export function HSL() { }
if (false) {
    /** @type {?} */
    HSL.prototype.h;
    /** @type {?} */
    HSL.prototype.s;
    /** @type {?} */
    HSL.prototype.l;
}
/**
 * @record
 */
export function HSLA() { }
if (false) {
    /** @type {?} */
    HSLA.prototype.a;
}
/**
 * @record
 */
export function HSV() { }
if (false) {
    /** @type {?} */
    HSV.prototype.h;
    /** @type {?} */
    HSV.prototype.s;
    /** @type {?} */
    HSV.prototype.v;
}
/**
 * @record
 */
export function HSVA() { }
if (false) {
    /** @type {?} */
    HSVA.prototype.a;
}
/**
 * @record
 */
export function HSVAsource() { }
if (false) {
    /** @type {?} */
    HSVAsource.prototype.source;
}
/**
 * @record
 */
export function HSLAsource() { }
if (false) {
    /** @type {?} */
    HSLAsource.prototype.source;
}
/**
 * @record
 */
export function Color() { }
if (false) {
    /** @type {?} */
    Color.prototype.hex;
    /** @type {?} */
    Color.prototype.rgb;
    /** @type {?} */
    Color.prototype.hsl;
    /** @type {?} */
    Color.prototype.hsv;
    /** @type {?} */
    Color.prototype.oldHue;
    /** @type {?} */
    Color.prototype.source;
}
/**
 * @record
 */
export function Shape() { }
if (false) {
    /** @type {?} */
    Shape.prototype.color;
    /** @type {?} */
    Shape.prototype.title;
}
