// ngx-color/sketch/esm5/sketch-preset-colors.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
var SketchPresetColorsComponent = /** @class */ (function () {
    function SketchPresetColorsComponent() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
        this.swatchStyle = {
            borderRadius: '3px',
            boxShadow: 'inset 0 0 0 1px rgba(0,0,0,.15)',
        };
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    SketchPresetColorsComponent.prototype.handleClick = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var hex = _a.hex, $event = _a.$event;
        this.onClick.emit({ hex: hex, $event: $event });
    };
    /**
     * @param {?} val
     * @return {?}
     */
    SketchPresetColorsComponent.prototype.normalizeValue = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        if (typeof val === 'string') {
            return { color: val };
        }
        return val;
    };
    /**
     * @param {?} val
     * @return {?}
     */
    SketchPresetColorsComponent.prototype.focusStyle = /**
     * @param {?} val
     * @return {?}
     */
    function (val) {
        /** @type {?} */
        var c = this.normalizeValue(val);
        return {
            boxShadow: "inset 0 0 0 1px rgba(0,0,0,.15), 0 0 4px " + c.color,
        };
    };
    SketchPresetColorsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-sketch-preset-colors',
                    template: "\n  <div class=\"sketch-swatches\">\n    <div class=\"sketch-wrap\" *ngFor=\"let c of colors\">\n      <color-swatch\n        [color]=\"normalizeValue(c).color\"\n        [style]=\"swatchStyle\"\n        [focusStyle]=\"focusStyle(c)\"\n        (onClick)=\"handleClick($event)\"\n        (onHover)=\"onSwatchHover.emit($event)\"\n        class=\"swatch\"\n      ></color-swatch>\n    </div>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n    .sketch-swatches {\n      margin: 0px -10px;\n      padding: 10px 0px 0px 10px;\n      border-top: 1px solid rgb(238, 238, 238);\n      display: flex;\n      flex-wrap: wrap;\n      position: relative;\n    }\n    .sketch-wrap {\n      width: 16px;\n      height: 16px;\n      margin: 0px 10px 10px 0px;\n    }\n  "]
                }] }
    ];
    SketchPresetColorsComponent.propDecorators = {
        colors: [{ type: Input }],
        onClick: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return SketchPresetColorsComponent;
}());
export { SketchPresetColorsComponent };
if (false) {
    /** @type {?} */
    SketchPresetColorsComponent.prototype.colors;
    /** @type {?} */
    SketchPresetColorsComponent.prototype.onClick;
    /** @type {?} */
    SketchPresetColorsComponent.prototype.onSwatchHover;
    /** @type {?} */
    SketchPresetColorsComponent.prototype.swatchStyle;
}
