// ngx-color/esm5/color-wrap.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { Component, EventEmitter, Input, NgModule, Output, } from '@angular/core';
import { debounceTime, distinctUntilChanged } from 'rxjs/operators';
import { simpleCheckForValidColor, toState } from './helpers/color';
/**
 * @record
 */
export function ColorEvent() { }
if (false) {
    /** @type {?} */
    ColorEvent.prototype.$event;
    /** @type {?} */
    ColorEvent.prototype.color;
}
var ColorWrap = /** @class */ (function () {
    function ColorWrap() {
        this.className = '';
        this.color = {
            h: 250,
            s: 0.5,
            l: 0.2,
            a: 1,
        };
        this.onChange = new EventEmitter();
        this.onChangeComplete = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ColorWrap.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.changes = this.onChange.pipe(debounceTime(100), distinctUntilChanged())
            .subscribe(function (x) { return _this.onChangeComplete.emit(x); });
        this.setState(toState(this.color, 0));
        this.currentColor = this.hex;
    };
    /**
     * @return {?}
     */
    ColorWrap.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.setState(toState(this.color, this.oldHue));
    };
    /**
     * @return {?}
     */
    ColorWrap.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.changes.unsubscribe();
    };
    /**
     * @param {?} data
     * @return {?}
     */
    ColorWrap.prototype.setState = /**
     * @param {?} data
     * @return {?}
     */
    function (data) {
        this.oldHue = data.oldHue;
        this.hsl = data.hsl;
        this.hsv = data.hsv;
        this.rgb = data.rgb;
        this.hex = data.hex;
        this.source = data.source;
        this.afterValidChange();
    };
    /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    ColorWrap.prototype.handleChange = /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    function (data, $event) {
        /** @type {?} */
        var isValidColor = simpleCheckForValidColor(data);
        if (isValidColor) {
            /** @type {?} */
            var color = toState(data, data.h || this.oldHue);
            this.setState(color);
            this.onChange.emit({ color: color, $event: $event });
            this.afterValidChange();
        }
    };
    /** hook for components after a complete change */
    /**
     * hook for components after a complete change
     * @return {?}
     */
    ColorWrap.prototype.afterValidChange = /**
     * hook for components after a complete change
     * @return {?}
     */
    function () { };
    /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    ColorWrap.prototype.handleSwatchHover = /**
     * @param {?} data
     * @param {?} $event
     * @return {?}
     */
    function (data, $event) {
        /** @type {?} */
        var isValidColor = simpleCheckForValidColor(data);
        if (isValidColor) {
            /** @type {?} */
            var color = toState(data, data.h || this.oldHue);
            this.setState(color);
            this.onSwatchHover.emit({ color: color, $event: $event });
        }
    };
    ColorWrap.decorators = [
        { type: Component, args: [{
                    // create seletor base for test override property
                    selector: 'color-wrap',
                    template: ""
                }] }
    ];
    ColorWrap.propDecorators = {
        className: [{ type: Input }],
        color: [{ type: Input }],
        onChange: [{ type: Output }],
        onChangeComplete: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return ColorWrap;
}());
export { ColorWrap };
if (false) {
    /** @type {?} */
    ColorWrap.prototype.className;
    /** @type {?} */
    ColorWrap.prototype.color;
    /** @type {?} */
    ColorWrap.prototype.onChange;
    /** @type {?} */
    ColorWrap.prototype.onChangeComplete;
    /** @type {?} */
    ColorWrap.prototype.onSwatchHover;
    /** @type {?} */
    ColorWrap.prototype.oldHue;
    /** @type {?} */
    ColorWrap.prototype.hsl;
    /** @type {?} */
    ColorWrap.prototype.hsv;
    /** @type {?} */
    ColorWrap.prototype.rgb;
    /** @type {?} */
    ColorWrap.prototype.hex;
    /** @type {?} */
    ColorWrap.prototype.source;
    /** @type {?} */
    ColorWrap.prototype.currentColor;
    /** @type {?} */
    ColorWrap.prototype.changes;
}
var ColorWrapModule = /** @class */ (function () {
    function ColorWrapModule() {
    }
    ColorWrapModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [ColorWrap],
                    exports: [ColorWrap],
                    imports: [CommonModule],
                },] }
    ];
    return ColorWrapModule;
}());
export { ColorWrapModule };
