// ngx-color/esm5/coordinates.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { Directive, ElementRef, HostListener, NgModule, Output, } from '@angular/core';
import { Subject } from 'rxjs';
import { distinctUntilChanged } from 'rxjs/operators';
var CoordinatesDirective = /** @class */ (function () {
    function CoordinatesDirective(el) {
        this.el = el;
        this.coordinatesChange = new Subject();
        this.mousechange = new Subject();
        this.mouseListening = false;
    }
    /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    CoordinatesDirective.prototype.mousemove = /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    function ($event, x, y, isTouch) {
        if (isTouch === void 0) { isTouch = false; }
        if (this.mouseListening) {
            $event.preventDefault();
            this.mousechange.next({ $event: $event, x: x, y: y, isTouch: isTouch });
        }
    };
    /**
     * @return {?}
     */
    CoordinatesDirective.prototype.mouseup = /**
     * @return {?}
     */
    function () {
        this.mouseListening = false;
    };
    /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    CoordinatesDirective.prototype.mousedown = /**
     * @param {?} $event
     * @param {?} x
     * @param {?} y
     * @param {?=} isTouch
     * @return {?}
     */
    function ($event, x, y, isTouch) {
        if (isTouch === void 0) { isTouch = false; }
        $event.preventDefault();
        this.mouseListening = true;
        this.mousechange.next({ $event: $event, x: x, y: y, isTouch: isTouch });
    };
    /**
     * @return {?}
     */
    CoordinatesDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.sub = this.mousechange
            .pipe(
        // limit times it is updated for the same area
        distinctUntilChanged(function (p, q) { return p.x === q.x && p.y === q.y; }))
            .subscribe(function (n) { return _this.handleChange(n.x, n.y, n.$event, n.isTouch); });
    };
    /**
     * @return {?}
     */
    CoordinatesDirective.prototype.ngOnDestroy = /**
     * @return {?}
     */
    function () {
        this.sub.unsubscribe();
    };
    /**
     * @param {?} x
     * @param {?} y
     * @param {?} $event
     * @param {?} isTouch
     * @return {?}
     */
    CoordinatesDirective.prototype.handleChange = /**
     * @param {?} x
     * @param {?} y
     * @param {?} $event
     * @param {?} isTouch
     * @return {?}
     */
    function (x, y, $event, isTouch) {
        /** @type {?} */
        var containerWidth = this.el.nativeElement.clientWidth;
        /** @type {?} */
        var containerHeight = this.el.nativeElement.clientHeight;
        /** @type {?} */
        var left = x -
            (this.el.nativeElement.getBoundingClientRect().left + window.pageXOffset);
        /** @type {?} */
        var top = y - this.el.nativeElement.getBoundingClientRect().top;
        if (!isTouch) {
            top = top - window.pageYOffset;
        }
        this.coordinatesChange.next({
            x: x,
            y: y,
            top: top,
            left: left,
            containerWidth: containerWidth,
            containerHeight: containerHeight,
            $event: $event,
        });
    };
    CoordinatesDirective.decorators = [
        { type: Directive, args: [{ selector: '[ngx-color-coordinates]' },] }
    ];
    /** @nocollapse */
    CoordinatesDirective.ctorParameters = function () { return [
        { type: ElementRef }
    ]; };
    CoordinatesDirective.propDecorators = {
        coordinatesChange: [{ type: Output }],
        mousemove: [{ type: HostListener, args: ['window:mousemove', ['$event', '$event.pageX', '$event.pageY'],] }, { type: HostListener, args: ['window:touchmove', [
                        '$event',
                        '$event.touches[0].clientX',
                        '$event.touches[0].clientY',
                        'true',
                    ],] }],
        mouseup: [{ type: HostListener, args: ['window:mouseup',] }, { type: HostListener, args: ['window:touchend',] }],
        mousedown: [{ type: HostListener, args: ['mousedown', ['$event', '$event.pageX', '$event.pageY'],] }, { type: HostListener, args: ['touchstart', [
                        '$event',
                        '$event.touches[0].clientX',
                        '$event.touches[0].clientY',
                        'true',
                    ],] }]
    };
    return CoordinatesDirective;
}());
export { CoordinatesDirective };
if (false) {
    /** @type {?} */
    CoordinatesDirective.prototype.coordinatesChange;
    /** @type {?} */
    CoordinatesDirective.prototype.mousechange;
    /** @type {?} */
    CoordinatesDirective.prototype.mouseListening;
    /** @type {?} */
    CoordinatesDirective.prototype.sub;
    /** @type {?} */
    CoordinatesDirective.prototype.el;
}
var CoordinatesModule = /** @class */ (function () {
    function CoordinatesModule() {
    }
    CoordinatesModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [CoordinatesDirective],
                    exports: [CoordinatesDirective],
                },] }
    ];
    return CoordinatesModule;
}());
export { CoordinatesModule };
