// ngx-color/esm5/raised.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule } from '@angular/core';
var RaisedComponent = /** @class */ (function () {
    function RaisedComponent() {
        this.zDepth = 1;
        this.radius = 1;
        this.background = '#fff';
    }
    RaisedComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-raised',
                    template: "\n  <div class=\"raised-wrap\">\n    <div class=\"raised-bg zDepth-{{zDepth}}\" [style.background]=\"background\"></div>\n    <div class=\"raised-content\">\n      <ng-content></ng-content>\n    </div>\n  </div>\n  ",
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n    .raised-wrap {\n      position: relative;\n      display: inline-block;\n    }\n    .raised-bg {\n      position: absolute;\n      top: 0px;\n      right: 0px;\n      bottom: 0px;\n      left: 0px;\n    }\n    .raised-content {\n      position: relative;\n    }\n    .zDepth-0 {\n      box-shadow: none;\n    }\n    .zDepth-1 {\n      box-shadow: 0 2px 10px rgba(0,0,0,.12), 0 2px 5px rgba(0,0,0,.16);\n    }\n    .zDepth-2 {\n      box-shadow: 0 6px 20px rgba(0,0,0,.19), 0 8px 17px rgba(0,0,0,.2);\n    }\n    .zDepth-3 {\n      box-shadow: 0 17px 50px rgba(0,0,0,.19), 0 12px 15px rgba(0,0,0,.24);\n    }\n    .zDepth-4 {\n      box-shadow: 0 25px 55px rgba(0,0,0,.21), 0 16px 28px rgba(0,0,0,.22);\n    }\n    .zDepth-5 {\n      box-shadow: 0 40px 77px rgba(0,0,0,.22), 0 27px 24px rgba(0,0,0,.2);\n    }\n  "]
                }] }
    ];
    RaisedComponent.propDecorators = {
        zDepth: [{ type: Input }],
        radius: [{ type: Input }],
        background: [{ type: Input }]
    };
    return RaisedComponent;
}());
export { RaisedComponent };
if (false) {
    /** @type {?} */
    RaisedComponent.prototype.zDepth;
    /** @type {?} */
    RaisedComponent.prototype.radius;
    /** @type {?} */
    RaisedComponent.prototype.background;
}
var RaisedModule = /** @class */ (function () {
    function RaisedModule() {
    }
    RaisedModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [RaisedComponent],
                    exports: [RaisedComponent],
                    imports: [CommonModule],
                },] }
    ];
    return RaisedModule;
}());
export { RaisedModule };
