// ngx-color/esm5/saturation.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, EventEmitter, Input, NgModule, Output, } from '@angular/core';
import { CoordinatesModule } from './coordinates.directive';
var SaturationComponent = /** @class */ (function () {
    function SaturationComponent() {
        this.onChange = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SaturationComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.background = "hsl(" + this.hsl.h + ", 100%, 50%)";
        this.pointerTop = -(this.hsv.v * 100) + 1 + 100 + '%';
        this.pointerLeft = this.hsv.s * 100 + '%';
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    SaturationComponent.prototype.handleChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var top = _a.top, left = _a.left, containerHeight = _a.containerHeight, containerWidth = _a.containerWidth, $event = _a.$event;
        if (left < 0) {
            left = 0;
        }
        else if (left > containerWidth) {
            left = containerWidth;
        }
        else if (top < 0) {
            top = 0;
        }
        else if (top > containerHeight) {
            top = containerHeight;
        }
        /** @type {?} */
        var saturation = left / containerWidth;
        /** @type {?} */
        var bright = -(top / containerHeight) + 1;
        bright = bright > 0 ? bright : 0;
        bright = bright > 1 ? 1 : bright;
        /** @type {?} */
        var data = {
            h: this.hsl.h,
            s: saturation,
            v: bright,
            a: this.hsl.a,
            source: 'hsva',
        };
        this.onChange.emit({ data: data, $event: $event });
    };
    SaturationComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-saturation',
                    template: "\n  <div class=\"color-saturation\" ngx-color-coordinates (coordinatesChange)=\"handleChange($event)\" [style.background]=\"background\">\n    <div class=\"saturation-white\">\n      <div class=\"saturation-black\"></div>\n      <div class=\"saturation-pointer\" [ngStyle]=\"pointer\" [style.top]=\"pointerTop\" [style.left]=\"pointerLeft\">\n        <div class=\"saturation-circle\" [ngStyle]=\"circle\"></div>\n      </div>\n    </div>\n  </div>\n  ",
                    preserveWhitespaces: false,
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    styles: ["\n    .saturation-white {\n      background: linear-gradient(to right, #fff, rgba(255,255,255,0));\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .saturation-black {\n      background: linear-gradient(to top, #000, rgba(0,0,0,0));\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .color-saturation {\n      position: absolute;\n      top: 0;\n      bottom: 0;\n      left: 0;\n      right: 0;\n    }\n    .saturation-pointer {\n      position: absolute;\n      cursor: default;\n    }\n    .saturation-circle {\n      width: 4px;\n      height: 4px;\n      box-shadow: 0 0 0 1.5px #fff, inset 0 0 1px 1px rgba(0,0,0,.3), 0 0 1px 2px rgba(0,0,0,.4);\n      border-radius: 50%;\n      cursor: hand;\n      transform: translate(-2px, -4px);\n    }\n  "]
                }] }
    ];
    SaturationComponent.propDecorators = {
        hsl: [{ type: Input }],
        hsv: [{ type: Input }],
        radius: [{ type: Input }],
        pointer: [{ type: Input }],
        circle: [{ type: Input }],
        onChange: [{ type: Output }]
    };
    return SaturationComponent;
}());
export { SaturationComponent };
if (false) {
    /** @type {?} */
    SaturationComponent.prototype.hsl;
    /** @type {?} */
    SaturationComponent.prototype.hsv;
    /** @type {?} */
    SaturationComponent.prototype.radius;
    /** @type {?} */
    SaturationComponent.prototype.pointer;
    /** @type {?} */
    SaturationComponent.prototype.circle;
    /** @type {?} */
    SaturationComponent.prototype.onChange;
    /** @type {?} */
    SaturationComponent.prototype.background;
    /** @type {?} */
    SaturationComponent.prototype.pointerTop;
    /** @type {?} */
    SaturationComponent.prototype.pointerLeft;
}
var SaturationModule = /** @class */ (function () {
    function SaturationModule() {
    }
    SaturationModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [SaturationComponent],
                    exports: [SaturationComponent],
                    imports: [CommonModule, CoordinatesModule],
                },] }
    ];
    return SaturationModule;
}());
export { SaturationModule };
