// ngx-color/esm5/swatch.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, EventEmitter, Input, NgModule, Output, } from '@angular/core';
import { CheckboardModule } from './checkboard.component';
var SwatchComponent = /** @class */ (function () {
    function SwatchComponent() {
        this.style = {};
        this.focusStyle = {};
        this.onClick = new EventEmitter();
        this.onHover = new EventEmitter();
        this.divStyles = {};
        this.focusStyles = {};
        this.inFocus = false;
    }
    /**
     * @return {?}
     */
    SwatchComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.divStyles = tslib_1.__assign({ background: this.color, height: '100%', width: '100%', cursor: 'pointer', position: 'relative', outline: 'none' }, this.style);
        this.focusStyles = tslib_1.__assign({}, this.divStyles, this.focusStyle);
    };
    /**
     * @return {?}
     */
    SwatchComponent.prototype.activeStyles = /**
     * @return {?}
     */
    function () {
        return this.focus || this.inFocus ? this.focusStyles : this.divStyles;
    };
    /**
     * @return {?}
     */
    SwatchComponent.prototype.handleFocusOut = /**
     * @return {?}
     */
    function () {
        this.inFocus = false;
    };
    /**
     * @return {?}
     */
    SwatchComponent.prototype.handleFocus = /**
     * @return {?}
     */
    function () {
        this.inFocus = true;
    };
    /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    SwatchComponent.prototype.handleHover = /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    function (hex, $event) {
        this.onHover.emit({ hex: hex, $event: $event });
    };
    /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    SwatchComponent.prototype.handleClick = /**
     * @param {?} hex
     * @param {?} $event
     * @return {?}
     */
    function (hex, $event) {
        this.onClick.emit({ hex: hex, $event: $event });
    };
    SwatchComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-swatch',
                    template: "\n  <div class=\"swatch\"\n    [ngStyle]=\"activeStyles()\" [attr.title]=\"color\"\n    (click)=\"handleClick(color, $event)\"\n    (keydown.enter)=\"handleClick(color, $event)\"\n    (focus)=\"handleFocus()\"\n    (focusout)=\"handleFocusOut()\"\n    (mouseover)=\"handleHover(color, $event)\"\n    tabindex=\"0\"\n  >\n    <ng-content></ng-content>\n    <color-checkboard\n      *ngIf=\"color === 'transparent'\"\n      boxShadow=\"inset 0 0 0 1px rgba(0,0,0,0.1)\"\n    ></color-checkboard>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false
                }] }
    ];
    SwatchComponent.propDecorators = {
        color: [{ type: Input }],
        style: [{ type: Input }],
        focusStyle: [{ type: Input }],
        focus: [{ type: Input }],
        onClick: [{ type: Output }],
        onHover: [{ type: Output }]
    };
    return SwatchComponent;
}());
export { SwatchComponent };
if (false) {
    /** @type {?} */
    SwatchComponent.prototype.color;
    /** @type {?} */
    SwatchComponent.prototype.style;
    /** @type {?} */
    SwatchComponent.prototype.focusStyle;
    /** @type {?} */
    SwatchComponent.prototype.focus;
    /** @type {?} */
    SwatchComponent.prototype.onClick;
    /** @type {?} */
    SwatchComponent.prototype.onHover;
    /** @type {?} */
    SwatchComponent.prototype.divStyles;
    /** @type {?} */
    SwatchComponent.prototype.focusStyles;
    /** @type {?} */
    SwatchComponent.prototype.inFocus;
}
var SwatchModule = /** @class */ (function () {
    function SwatchModule() {
    }
    SwatchModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [SwatchComponent],
                    exports: [SwatchComponent],
                    imports: [CommonModule, CheckboardModule],
                },] }
    ];
    return SwatchModule;
}());
export { SwatchModule };
