// ngx-color/photoshop/esm5/photoshop-previews.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, Input, } from '@angular/core';
var PhotoshopPreviewsComponent = /** @class */ (function () {
    function PhotoshopPreviewsComponent() {
        this.currentColor = '';
        this.backgroundNew = '';
    }
    /**
     * @return {?}
     */
    PhotoshopPreviewsComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.backgroundNew = "rgb(" + this.rgb.r + "," + this.rgb.g + ", " + this.rgb.b + ")";
    };
    PhotoshopPreviewsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-photoshop-previews',
                    template: "\n  <div>\n    <div class=\"photoshop-label\">new</div>\n    <div class=\"photoshop-swatches\">\n      <div class=\"photoshop-new\" [style.background]=\"backgroundNew\"></div>\n      <div class=\"photoshop-current\" [style.background]=\"currentColor\"></div>\n    </div>\n    <div class=\"photoshop-label\">current</div>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n  .photoshop-swatches {\n    border: 1px solid #B3B3B3;\n    border-bottom: 1px solid #F0F0F0;\n    margin-bottom: 2px;\n    margin-top: 1px;\n  }\n  .photoshop-new {\n    height: 34px;\n    box-shadow: inset 1px 0 0 #000, inset -1px 0 0 #000, inset 0 1px 0 #000;\n  }\n  .photoshop-current {\n    height: 34px;\n    box-shadow: inset 1px 0 0 #000, inset -1px 0 0 #000, inset 0 -1px 0 #000;\n  }\n  .photoshop-label {\n    font-size: 14px;\n    color: #000;\n    text-align: center;\n  }\n  "]
                }] }
    ];
    PhotoshopPreviewsComponent.propDecorators = {
        rgb: [{ type: Input }],
        currentColor: [{ type: Input }]
    };
    return PhotoshopPreviewsComponent;
}());
export { PhotoshopPreviewsComponent };
if (false) {
    /** @type {?} */
    PhotoshopPreviewsComponent.prototype.rgb;
    /** @type {?} */
    PhotoshopPreviewsComponent.prototype.currentColor;
    /** @type {?} */
    PhotoshopPreviewsComponent.prototype.backgroundNew;
}
