// ngx-color/swatches/esm5/swatches-color.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { getContrastingColor } from 'ngx-color';
var SwatchesColorComponent = /** @class */ (function () {
    function SwatchesColorComponent() {
        this.first = false;
        this.last = false;
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
        this.getContrastingColor = getContrastingColor;
        this.colorStyle = {
            width: '40px',
            height: '24px',
            cursor: 'pointer',
            marginBottom: '1px',
        };
        this.focusStyle = {};
    }
    /**
     * @return {?}
     */
    SwatchesColorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.colorStyle.background = this.color;
        this.focusStyle.boxShadow = "0 0 4px " + this.color;
        if (this.first) {
            this.colorStyle.borderRadius = '2px 2px 0 0';
        }
        if (this.last) {
            this.colorStyle.borderRadius = '0 0 2px 2px';
        }
        if (this.color === '#FFFFFF') {
            this.colorStyle.boxShadow = 'inset 0 0 0 1px #ddd';
        }
    };
    /**
     * @param {?} $event
     * @return {?}
     */
    SwatchesColorComponent.prototype.handleClick = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.onClick.emit({
            data: {
                hex: this.color,
                source: 'hex',
            },
            $event: $event,
        });
    };
    SwatchesColorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-swatches-color',
                    template: "\n  <color-swatch [color]=\"color\" [style]=\"colorStyle\" [focusStyle]=\"focusStyle\"\n    [class.first]=\"first\" [class.last]=\"last\"\n    (click)=\"handleClick($event)\" (onHover)=\"onSwatchHover.emit($event)\">\n    <div class=\"swatch-check\" *ngIf=\"active\" [class.first]=\"first\" [class.last]=\"last\">\n      <svg style=\"width: 24px; height: 24px;\" viewBox=\"0 0 24 24\" [style.fill]=\"getContrastingColor(color)\">\n        <path d=\"M21,7L9,19L3.5,13.5L4.91,12.09L9,16.17L19.59,5.59L21,7Z\" />\n      </svg>\n    </div>\n  </color-swatch>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n    .swatches-group {\n      padding-bottom: 10px;\n      width: 40px;\n      float: left;\n      margin-right: 10px;\n    }\n    .swatch-check {\n      margin-left: 8px;\n    }\n  "]
                }] }
    ];
    SwatchesColorComponent.propDecorators = {
        color: [{ type: Input }],
        first: [{ type: Input }],
        last: [{ type: Input }],
        active: [{ type: Input }],
        onClick: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return SwatchesColorComponent;
}());
export { SwatchesColorComponent };
if (false) {
    /** @type {?} */
    SwatchesColorComponent.prototype.color;
    /** @type {?} */
    SwatchesColorComponent.prototype.first;
    /** @type {?} */
    SwatchesColorComponent.prototype.last;
    /** @type {?} */
    SwatchesColorComponent.prototype.active;
    /** @type {?} */
    SwatchesColorComponent.prototype.onClick;
    /** @type {?} */
    SwatchesColorComponent.prototype.onSwatchHover;
    /** @type {?} */
    SwatchesColorComponent.prototype.getContrastingColor;
    /** @type {?} */
    SwatchesColorComponent.prototype.colorStyle;
    /** @type {?} */
    SwatchesColorComponent.prototype.focusStyle;
}
