// ngx-color/swatches/esm5/swatches-group.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
var SwatchesGroupComponent = /** @class */ (function () {
    function SwatchesGroupComponent() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
    }
    SwatchesGroupComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-swatches-group',
                    template: "\n  <div class=\"swatches-group\">\n    <color-swatches-color\n      *ngFor=\"let color of group; let idx = index;\"\n      [active]=\"color.toLowerCase() === active\"\n      [color]=\"color\"\n      [first]=\"idx === 0\"\n      [last]=\"idx === group.length - 1\"\n      (onClick)=\"onClick.emit($event)\"\n    >\n    </color-swatches-color>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n    .swatches-group {\n      padding-bottom: 10px;\n      width: 40px;\n      float: left;\n      margin-right: 10px;\n    }\n  "]
                }] }
    ];
    SwatchesGroupComponent.propDecorators = {
        group: [{ type: Input }],
        active: [{ type: Input }],
        onClick: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return SwatchesGroupComponent;
}());
export { SwatchesGroupComponent };
if (false) {
    /** @type {?} */
    SwatchesGroupComponent.prototype.group;
    /** @type {?} */
    SwatchesGroupComponent.prototype.active;
    /** @type {?} */
    SwatchesGroupComponent.prototype.onClick;
    /** @type {?} */
    SwatchesGroupComponent.prototype.onSwatchHover;
}
