// ngx-color/hue/esm5/hue-picker.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { CommonModule } from '@angular/common';
import { ChangeDetectionStrategy, Component, Input, NgModule, } from '@angular/core';
import { ColorWrap, HueModule } from 'ngx-color';
import { toState } from 'ngx-color';
var HuePickerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(HuePickerComponent, _super);
    function HuePickerComponent() {
        var _this = _super.call(this) || this;
        /**
         * Pixel value for picker width
         */
        _this.width = 316;
        /**
         * Pixel value for picker height
         */
        _this.height = 16;
        _this.radius = 2;
        _this.direction = 'horizontal';
        _this.pointer = {
            width: '18px',
            height: '18px',
            borderRadius: '50%',
            transform: 'translate(-9px, -2px)',
            backgroundColor: 'rgb(248, 248, 248)',
            boxShadow: '0 1px 4px 0 rgba(0, 0, 0, 0.37)',
        };
        return _this;
    }
    /**
     * @return {?}
     */
    HuePickerComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        if (this.direction === 'vertical') {
            this.pointer.transform = 'translate(-3px, -9px)';
        }
        this.setState(toState(this.color, this.oldHue));
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    HuePickerComponent.prototype.handlePickerChange = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var data = _a.data, $event = _a.$event;
        this.handleChange({ a: 1, h: data.h, l: 0.5, s: 1 }, $event);
    };
    HuePickerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-hue-picker',
                    template: "\n  <div class=\"hue-picker {{ className }}\"\n    [style.width.px]=\"width\" [style.height.px]=\"height\"\n  >\n    <color-hue [hsl]=\"hsl\" [pointer]=\"pointer\"\n      [direction]=\"direction\" [radius]=\"radius\"\n      (onChange)=\"handlePickerChange($event)\"\n    ></color-hue>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n    .hue-picker {\n      position: relative;\n    }\n  "]
                }] }
    ];
    /** @nocollapse */
    HuePickerComponent.ctorParameters = function () { return []; };
    HuePickerComponent.propDecorators = {
        width: [{ type: Input }],
        height: [{ type: Input }],
        radius: [{ type: Input }],
        direction: [{ type: Input }]
    };
    return HuePickerComponent;
}(ColorWrap));
export { HuePickerComponent };
if (false) {
    /**
     * Pixel value for picker width
     * @type {?}
     */
    HuePickerComponent.prototype.width;
    /**
     * Pixel value for picker height
     * @type {?}
     */
    HuePickerComponent.prototype.height;
    /** @type {?} */
    HuePickerComponent.prototype.radius;
    /** @type {?} */
    HuePickerComponent.prototype.direction;
    /** @type {?} */
    HuePickerComponent.prototype.pointer;
}
var ColorHueModule = /** @class */ (function () {
    function ColorHueModule() {
    }
    ColorHueModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [HuePickerComponent],
                    exports: [HuePickerComponent],
                    imports: [CommonModule, HueModule],
                },] }
    ];
    return ColorHueModule;
}());
export { ColorHueModule };
