// ngx-color/compact/esm5/compact-color.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
 */
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output, } from '@angular/core';
import { getContrastingColor } from 'ngx-color';
var CompactColorComponent = /** @class */ (function () {
    function CompactColorComponent() {
        this.onClick = new EventEmitter();
        this.onSwatchHover = new EventEmitter();
        this.swatchStyle = {
            width: '15px',
            height: '15px',
            float: 'left',
            marginRight: '5px',
            marginBottom: '5px',
            position: 'relative',
            cursor: 'pointer',
        };
        this.swatchFocus = {};
        this.getContrastingColor = getContrastingColor;
    }
    /**
     * @return {?}
     */
    CompactColorComponent.prototype.ngOnChanges = /**
     * @return {?}
     */
    function () {
        this.swatchStyle.background = this.color;
        this.swatchFocus.boxShadow = "0 0 4px " + this.color;
        if (this.color.toLowerCase() === '#ffffff') {
            this.swatchStyle.boxShadow = 'inset 0 0 0 1px #ddd';
        }
    };
    /**
     * @param {?} __0
     * @return {?}
     */
    CompactColorComponent.prototype.handleClick = /**
     * @param {?} __0
     * @return {?}
     */
    function (_a) {
        var hex = _a.hex, $event = _a.$event;
        this.onClick.emit({ hex: hex, $event: $event });
    };
    CompactColorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'color-compact-color',
                    template: "\n  <div class=\"compact-color\">\n    <color-swatch class=\"swatch\"\n      [color]=\"color\" [style]=\"swatchStyle\"\n      [focusStyle]=\"swatchFocus\"\n      (onClick)=\"handleClick($event)\" (onHover)=\"onSwatchHover.emit($event)\"\n      >\n      <div class=\"compact-dot\"\n        [class.active]=\"active\" [style.background]=\"getContrastingColor(color)\"\n      ></div>\n    </color-swatch>\n  </div>\n  ",
                    changeDetection: ChangeDetectionStrategy.OnPush,
                    preserveWhitespaces: false,
                    styles: ["\n  .compact-dot {\n    position: absolute;\n    top: 5px;\n    right: 5px;\n    bottom: 5px;\n    left: 5px;\n    border-radius: 50%;\n    opacity: 0;\n  }\n  .compact-dot.active {\n    opacity: 1;\n  }\n  "]
                }] }
    ];
    CompactColorComponent.propDecorators = {
        color: [{ type: Input }],
        active: [{ type: Input }],
        onClick: [{ type: Output }],
        onSwatchHover: [{ type: Output }]
    };
    return CompactColorComponent;
}());
export { CompactColorComponent };
if (false) {
    /** @type {?} */
    CompactColorComponent.prototype.color;
    /** @type {?} */
    CompactColorComponent.prototype.active;
    /** @type {?} */
    CompactColorComponent.prototype.onClick;
    /** @type {?} */
    CompactColorComponent.prototype.onSwatchHover;
    /** @type {?} */
    CompactColorComponent.prototype.swatchStyle;
    /** @type {?} */
    CompactColorComponent.prototype.swatchFocus;
    /** @type {?} */
    CompactColorComponent.prototype.getContrastingColor;
}
